plugins/destination/s3/client/streamingbatch.go

```diff
@@ -0,0 +1,408 @@
+// streamingbatch.go copied from file plugin
+package client
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/apache/arrow/go/v13/arrow"
+	"github.com/apache/arrow/go/v13/arrow/array"
+	"github.com/apache/arrow/go/v13/arrow/util"
+	"github.com/cloudquery/plugin-sdk/v4/message"
+	"github.com/cloudquery/plugin-sdk/v4/schema"
+	"github.com/rs/zerolog"
+)
+
+const (
+	defaultBatchTimeoutSeconds = 20
+	defaultBatchSize           = 10000
+	defaultBatchSizeBytes      = 5 * 1024 * 1024 // 5 MiB
+)
+
+type StreamingBatchWriterClient interface {
+	MigrateTables(context.Context, []*message.MigrateTable) error
+	DeleteStale(context.Context, []*message.DeleteStale) error
+
+	OpenTable(ctx context.Context, sourceName string, table *schema.Table, syncTime time.Time) (any, error)
+	WriteTableStream(ctx context.Context, handle any, upsert bool, msgs []*message.Insert) error
+	CloseTable(ctx context.Context, handle any) error
+}
+
+type StreamingBatchWriter struct {
+	client           StreamingBatchWriterClient
+	workers          map[string]*worker
+	workersLock      *sync.RWMutex
+	workersWaitGroup *sync.WaitGroup
+
+	migrateTableLock     *sync.Mutex
+	migrateTableMessages []*message.MigrateTable
+	deleteStaleLock      *sync.Mutex
+	deleteStaleMessages  []*message.DeleteStale
+
+	logger         zerolog.Logger
+	batchTimeout   time.Duration
+	batchSize      int
+	batchSizeBytes int
+}
```

Comment:
if you're going to use `StreamingBatchWriter` only by pointer there's no need to define sync primitives as pointers inside (define as value & don't alloc explicitly):
```suggestion
type StreamingBatchWriter struct {
	client           StreamingBatchWriterClient
	workers          map[string]*worker
	workersLock      sync.RWMutex
	workersWaitGroup sync.WaitGroup

	migrateTableLock     sync.Mutex
	migrateTableMessages []*message.MigrateTable
	deleteStaleLock      sync.Mutex
	deleteStaleMessages  []*message.DeleteStale

	logger         zerolog.Logger
	batchTimeout   time.Duration
	batchSize      int
	batchSizeBytes int
}
```

this allows for the init:
```go
c := &StreamingBatchWriter{
	client:           client,
	workers:          make(map[string]*worker),
	logger:           zerolog.Nop(),
	batchTimeout:     defaultBatchTimeoutSeconds * time.Second,
	batchSize:        defaultBatchSize,
	batchSizeBytes:   defaultBatchSizeBytes,
}
```